File: lifebook/android/app/src/main/java/com/beyondpandora/hughmanatee/ui/session/SessionDetailViewModel.kt
package com.beyondpandora.hughmanatee.ui.session

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.beyondpandora.hughmanatee.data.local.HughDatabase
import com.beyondpandora.hughmanatee.data.local.entities.SessionEntity
import com.beyondpandora.hughmanatee.data.local.entities.TurnEntity
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class SessionDetailUiState(
    val session: SessionEntity? = null,
    val turns: List<TurnEntity> = emptyList(),
    val isLoading: Boolean = true,
)

class SessionDetailViewModel(
    private val db: HughDatabase,
    private val sessionId: String,
) : ViewModel() {

    private val sessionRepo = com.beyondpandora.hughmanatee.data.repository.SessionRepository(db)

    private val _state = MutableStateFlow(SessionDetailUiState())
    val state: StateFlow<SessionDetailUiState> = _state.asStateFlow()

    init {
        viewModelScope.launch {
            val session = sessionRepo.getSession(sessionId)
            val turns = sessionRepo.getTurns(sessionId)
            _state.value = SessionDetailUiState(
                session = session,
                turns = turns,
                isLoading = false,
            )
        }
    }
}

File: lifebook/ios-native/HughManatee/ViewModels/SessionDetailViewModel.swift
import SwiftUI

@Observable
final class SessionDetailViewModel {
    var session: SessionRecord?
    var turns: [TurnRecord] = []
    var isLoading = true

    private let db = DatabaseService.shared

    init(sessionId: String) {
        Task {
            do {
                session = try await db.reader.read { db in
                    try SessionRecord.fetchOne(db, key: sessionId)
                }
                turns = try await db.getTurns(sessionId: sessionId)
                isLoading = false
            } catch {
                isLoading = false
            }
        }
    }
}
